<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <title>API Design - IBDCluster</title> <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"> <link rel="stylesheet" href="/assets/css/just-the-docs-default.css"> <script type="text/javascript" src="/assets/js/vendor/lunr.min.js"></script> <script type="text/javascript" src="/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.8.0 --> <title>API Design | IBDCluster</title> <meta name="generator" content="Jekyll v3.9.2" /> <meta property="og:title" content="API Design" /> <meta property="og:locale" content="en_US" /> <meta name="description" content="A cli tool to analyze ibd sharing for a specific locus for multiple phenotypes" /> <meta property="og:description" content="A cli tool to analyze ibd sharing for a specific locus for multiple phenotypes" /> <link rel="canonical" href="http://localhost:4000/documentation/api_design.html" /> <meta property="og:url" content="http://localhost:4000/documentation/api_design.html" /> <meta property="og:site_name" content="IBDCluster" /> <meta property="og:type" content="website" /> <meta name="twitter:card" content="summary" /> <meta property="twitter:title" content="API Design" /> <script type="application/ld+json"> {"@context":"https://schema.org","@type":"WebPage","description":"A cli tool to analyze ibd sharing for a specific locus for multiple phenotypes","headline":"API Design","url":"http://localhost:4000/documentation/api_design.html"}</script> <!-- End Jekyll SEO tag --> </head> <body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="svg-link" viewBox="0 0 24 24"> <title>Link</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"> <path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"> <title>Search</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"> <title>Menu</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"> <line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"> <title>Expand</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"> <polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"> <title>Document</title> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"> <path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> <!-- Feather. MIT License: https://github.com/feathericons/feather/blob/master/LICENSE --> <symbol id="svg-external-link" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-external-link"> <title id="svg-external-link-title">(external link)</title> <path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path><polyline points="15 3 21 3 21 9"></polyline><line x1="10" y1="14" x2="21" y2="3"></line> </symbol> </svg> <div class="side-bar"> <div class="site-header"> <a href="/" class="site-title lh-tight"> IBDCluster </a> <a href="#" id="menu-button" class="site-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-menu"></use></svg> </a> </div> <nav role="navigation" aria-label="Main" id="site-nav" class="site-nav"> <ul class="nav-list"><li class="nav-list-item"><a href="/" class="nav-list-link">Home</a></li><li class="nav-list-item active"><a href="#" class="nav-list-expander"><svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg></a><a href="/documentation/" class="nav-list-link">Documentation</a><ul class="nav-list "><li class="nav-list-item "><a href="#" class="nav-list-expander"><svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg></a><a href="/documentation/installation.html" class="nav-list-link">Installation</a><ul class="nav-list"><li class="nav-list-item "> <a href="/documentation/github.html" class="nav-list-link">Github Installation</a> </li><li class="nav-list-item "> <a href="/documentation/pip.html" class="nav-list-link">Pip Installation</a> </li><li class="nav-list-item "> <a href="/documentation/docker.html" class="nav-list-link">Docker Installation</a> </li></ul></li><li class="nav-list-item "><a href="/documentation/inputs.html" class="nav-list-link">Inputs</a></li><li class="nav-list-item active"><a href="#" class="nav-list-expander"><svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg></a><a href="/documentation/customization.html" class="nav-list-link">Customization</a><ul class="nav-list"><li class="nav-list-item active"> <a href="/documentation/api_design.html" class="nav-list-link active">API Design</a> </li><li class="nav-list-item "> <a href="/documentation/stock_plugins.html" class="nav-list-link">Stock Plugins</a> </li><li class="nav-list-item "> <a href="/documentation/design_your_own_plugin.html" class="nav-list-link">Custom Pluging Design</a> </li></ul></li></ul></li><li class="nav-list-item"><a href="#" class="nav-list-expander"><svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg></a><a href="/examples/" class="nav-list-link">Example Scripts</a><ul class="nav-list "></ul></li><li class="nav-list-item"><a href="/changelog.html" class="nav-list-link">Changelog</a></li></ul> </nav> <footer class="site-footer"> This site uses <a href="https://github.com/just-the-docs/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll. </footer> </div> <div class="main" id="top"> <div id="main-header" class="main-header"> <div class="search"> <div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search IBDCluster" aria-label="Search IBDCluster" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label> </div> <div id="search-results" class="search-results"></div> </div> <nav aria-label="Auxiliary" class="aux-nav"> <ul class="aux-nav-list"> <li class="aux-nav-list-item"> <a href="https://github.com/jtb324/IBDCluster" class="site-button" target="_blank" rel="noopener noreferrer" > IBDCluster on Github </a> </li> </ul> </nav> </div> <div id="main-content-wrap" class="main-content-wrap"> <nav aria-label="Breadcrumb" class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="/documentation/">Documentation</a></li> <li class="breadcrumb-nav-list-item"><a href="/documentation/customization.html">Customization</a></li> <li class="breadcrumb-nav-list-item"><span>API Design</span></li> </ol> </nav> <div id="main-content" class="main-content" role="main"> <h1 id="api-design"> <a href="#api-design" class="anchor-heading" aria-labelledby="api-design"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> API Design: </h1> <hr /> <p>The api is designed so that the user can access the following models as well as the output path for the program:</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">@</span><span class="n">dataclass</span>
<span class="k">class</span> <span class="nc">DataHolder</span><span class="p">:</span>
    <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">chromosome</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">affected_inds</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
    <span class="n">phenotype_prevalence</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span>
    <span class="n">phenotype_cols</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
    <span class="n">ibd_program</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">phenotype_description</span><span class="p">:</span> <span class="bp">None</span> <span class="o">|</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="bp">None</span>

<span class="o">@</span><span class="n">dataclass</span>
<span class="k">class</span> <span class="nc">Pairs</span><span class="p">:</span>
    <span class="n">pair_1</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">phase_1</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">pair_2</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">phase_2</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">chromosome</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">segment_start</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">segment_end</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">length</span><span class="p">:</span> <span class="nb">float</span>

    <span class="k">def</span> <span class="nf">form_id_str</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="s">"""Method that will return the pair ids and phases in a formated string"""</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s">"</span><span class="si">{</span><span class="bp">self</span><span class="p">.</span><span class="n">pair_1</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="bp">self</span><span class="p">.</span><span class="n">pair_2</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="bp">self</span><span class="p">.</span><span class="n">phase_1</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="bp">self</span><span class="p">.</span><span class="n">phase_2</span><span class="si">}</span><span class="s">"</span>

    <span class="k">def</span> <span class="nf">form_segment_info_str</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="s">"""Method that will return a string with all of the segment info"""</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s">"</span><span class="si">{</span><span class="bp">self</span><span class="p">.</span><span class="n">segment_start</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="bp">self</span><span class="p">.</span><span class="n">segment_end</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="bp">self</span><span class="p">.</span><span class="n">length</span><span class="si">}</span><span class="se">\n</span><span class="s">"</span>

<span class="o">@</span><span class="n">dataclass</span>
<span class="k">class</span> <span class="nc">Network</span><span class="p">:</span>
    <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">gene_chr</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">network_id</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">pairs</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Pairs</span><span class="p">]</span> <span class="o">=</span> <span class="n">field</span><span class="p">(</span><span class="n">default_factory</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>
    <span class="n">iids</span><span class="p">:</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">field</span><span class="p">(</span><span class="n">default_factory</span><span class="o">=</span><span class="nb">set</span><span class="p">)</span>
    <span class="n">haplotypes</span><span class="p">:</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">field</span><span class="p">(</span><span class="n">default_factory</span><span class="o">=</span><span class="nb">set</span><span class="p">)</span>
    <span class="n">connections</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="n">field</span><span class="p">(</span><span class="n">default_factory</span><span class="o">=</span><span class="nb">dict</span><span class="p">)</span>
    <span class="n">pvalues</span><span class="p">:</span> <span class="bp">None</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="bp">None</span>
</code></pre></div></div> <p>The api will supply three values: “data”, “network”, “output” which are represented by the DataHolder model, the Network model, and the gene output path respectively. You can access each of these values through the kwargs. These attributes and methods of each model are described below.</p> <h2 id="dataholder"> <a href="#dataholder" class="anchor-heading" aria-labelledby="dataholder"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> DataHolder: </h2> <hr /> <p>The attributes for the DataHolder model are described below:</p> <ul> <li> <p><span style="color: #F0FF00"><strong>gene_name</strong></span> - string that contains the gene name that was passed to the program in the gene_info file</p> </li> <li> <p><span style="color: #F0FF00"><strong>chromosome</strong></span> - integer number for the chromosome of interest</p> </li> <li> <p><span style="color: #F0FF00"><strong>affected_inds</strong></span> - dictionary where the keys are a string for each PheCode and the values are a list of IDs that are affected by the phenotype.</p> </li> <li> <p><span style="color: #F0FF00"><strong>phenotype_prevalence</strong></span> - dictionary where the keys are a PheCode string and the values are the percentages of individuals that are affected by the phenotype in the population you provided.</p> </li> <li> <p><span style="color: #F0FF00"><strong>phenotype_cols</strong></span> - list of PheCode strings in the provided PheCode matrix.</p> </li> <li> <p><span style="color: #F0FF00"><strong>ibd_program</strong></span> - this is a string that gives the name of the ibd program being used.</p> </li> <li> <p><span style="color: #F0FF00"><strong>phenotype_description</strong></span> - This attribute is either none if the user didn’t provide a file that has descriptions of each PheCode or it is a dictionary where the keys are PheCode strings and the values are the description of each PheCode.</p> </li> </ul> <h2 id="pairs"> <a href="#pairs" class="anchor-heading" aria-labelledby="pairs"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Pairs: </h2> <hr /> <p>The attributes for the Pairs model are described below:</p> <ul> <li> <p><span style="color: #F0FF00"><strong>pair_1</strong></span> - string that contains the ID of the first person in the pair that shares an IBD segment.</p> </li> <li> <p><span style="color: #F0FF00"><strong>phase_1</strong></span> - string that contains ID and the phase number of pair_1’s haplotype which contains the shared IBD segment. This values is the pair ID concatenated with the phase value. Ex: ID1 becomes ID1.1 for the first haplotype (Different ibd detection programs use different numbers to represent haplotypes. The example above is based on hap-IBD).</p> </li> <li> <p><span style="color: #F0FF00"><strong>pair_2</strong></span> - string that contains the ID of the second person in the pair that shares an IBD segment.</p> </li> <li> <p><span style="color: #F0FF00"><strong>phase_2</strong></span> - string that contains the ID and the phase number of pair_2’s haplotype which contains the shared IBD segment. This values is the pair ID concatenated with the phase value. Ex: ID2 becomes ID2.2 for the second haplotype (Different ibd detection programs use different numbers to represent haplotypes. The example above is based on hap-IBD).</p> </li> <li> <p><span style="color: #F0FF00"><strong>chromosome</strong></span> - integer number of the chromosome that the shared segment is on.</p> </li> <li> <p><span style="color: #F0FF00"><strong>segment_start</strong></span> - integer that has the start base position of the shared segment.</p> </li> <li> <p><span style="color: #F0FF00"><strong>segment_end</strong></span> - integer that has the end base position of the shared segment.</p> </li> <li> <p><span style="color: #F0FF00"><strong>length</strong></span> - float value that has the total length of the segments in centimorgans.</p> </li> </ul> <p>The methods for the Pairs model are described below:</p> <ul> <li> <p><span style="color: #00FFE0"><strong>form_id_str</strong></span> - method that creates a string with tab separated values that has the pair_1 ID string, the pair_2 ID string, the phase_1 string, and the phase_2 string, respectively.</p> </li> <li> <p><span style="color: #00FFE0"><strong>form_segment_info_str</strong></span> - method that creates a string with tab separated values that has the segment start position, the segment end position, and the length of the segment, respectively.</p> </li> </ul> <h2 id="network"> <a href="#network" class="anchor-heading" aria-labelledby="network"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Network: </h2> <hr /> <p>The attributes for the Network model are described below:</p> <ul> <li> <p><span style="color: #F0FF00"><strong>gene_name</strong></span> - string that contains the gene name that was passed to the program in the gene_info file.</p> </li> <li> <p><span style="color: #F0FF00"><strong>gene_chr</strong></span> - integer that has the chromosome number that the gene is on.</p> </li> <li> <p><span style="color: #F0FF00"><strong>network_id</strong></span> - integer that has the id for the network in the analysis.</p> </li> <li> <p><span style="color: #F0FF00"><strong>pairs</strong></span> - list of Pair objects that have information about the ibd segment that each id pair shares.</p> </li> <li> <p><span style="color: #F0FF00"><strong>iids</strong></span> - a set that has all the unique iid strings that are in the network.</p> </li> <li> <p><span style="color: #F0FF00"><strong>haplotypes</strong></span> - a set that has all the unique haplotypes in the network. The unique haplotypes are just the ID with the phase information appended to it. Ex: ID1 becomes ID1.1. This attribute allows us to differentiate between which haplotype is in the network.</p> </li> <li> <p><span style="color: #F0FF00"><strong>connections</strong></span> - a dictionary where the keys are the individual grid strings and the values are the number of other individuals that the ID is connected to.</p> </li> <li> <p><span style="color: #F0FF00"><strong>pvalues</strong></span> - This value is either None or it is a string of tab separated values where the first three values are the most significant p-value, the corresponding PheCode, and then the corresponding PheCode description respectively. The rest of the string are tab separated values for the p-value of each PheCode and how many individuals in the network are affected by the Phenotype. The default value for this attribute is None and the pvalue.py plugin actually adds the string to this attribute. The user could substitute there own statistically plugin if they wished and use this attribute, it just needs to meet the plugin requirements which are defined in the “Custom Pluging Design” section.</p> </li> </ul> <h2 id="output-path"> <a href="#output-path" class="anchor-heading" aria-labelledby="output-path"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Output Path: </h2> <hr /> <p>The api also supplies the gene output path. This output path is formed by affixing the name of the gene of interest with the output path provided by the user to the CLI. The plugin needs to create the path to write information to.</p> <p>for example if you were running the gene rbm20, then this value would be:</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/output_path_provided_by_user/rbm20/
</code></pre></div></div> </div> </div> <div class="search-overlay"></div> </div> </body> </html>
